server/treds_fsm.go

```diff
@@ -26,17 +29,50 @@ func (t TredsFsm) Apply(log *raft.Log) interface{} {
 	return commandReg.Execute(commandStringParts[1:], t.tredsStore)
 }
 
+type snapshot struct {
+	storageSnapshot []byte
+}
+
+func (s *snapshot) Persist(sink raft.SnapshotSink) error {
+	if _, err := sink.Write(s.storageSnapshot); err != nil {
+		return err
+	}
+	if err := sink.Close(); err != nil {
+		return fmt.Errorf("failed to close snapshot sink: %v", err)
+	}
+	return nil
+}
+
+func (s *snapshot) Release() {}
+
 func (t TredsFsm) Snapshot() (raft.FSMSnapshot, error) {
-	//TODO: implement snapshot creation
-	//This need to read the full im mem data, serialize it and write it to the snapshot
-	panic("implement me")
+	defer func(start time.Time) {
+		log.Println("snapshot created", "duration", time.Since(start).String())
+	}(time.Now())
+
+	storageSnapshot, err := t.tredsStore.Snapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	return &snapshot{
+		storageSnapshot: storageSnapshot,
+	}, nil
 }
 
-func (t TredsFsm) Restore(_ io.ReadCloser) error {
-	//TODO: implement snapshot creation
-	//This need to read from the snapshot, parse the commands,
-	//use the registry to retrieve the right command and execute it with the data
-	panic("implement me")
+func (t TredsFsm) Restore(old io.ReadCloser) error {
+	defer old.Close()
+	data, err := io.ReadAll(old)
+	if err != nil {
+		return err
+	}
+	ts := store.NewTredsStore()
+	err = ts.Restore(data)
+	if err != nil {
+		return err
+	}
+	t.tredsStore = ts
```

Comment:
I'm not sure where `Restore` could be called, but this is not thread safe. You need to make the `tredsStore` pointer atomic to make it safe.

Reply:
@dhiaayachi the system is running a single thread, is this required?